### ISSUE TITLE
Strict object literal checking with union type

### ISSUE THREAD

**Issue by Lenne231** [Microsoft/TypeScript#9470]
**TypeScript Version:**  1.8.0

Is this a bug or does it work as intended?

**Code**

```ts
type Test = { a: string; } | { b:number };

var t1 : Test = { a: '...', b: 3 }; // ok
var t2 : Test = { a: '...', b: '...' }; // ok but should be error!?
var t3 : Test = { a: '...', c: '...' }; // error

var a4 : { a: string; } = { a: '...', b: '...' }; // error
```

**Expected behavior:**
`var t2 : Test = { a: '...', b: '...' };` should throw a compiler error as well because `b` is of type `string` and not `number`.

**Actual behavior:** 
There is no compiler error.

**Comment by DanielRosenwasser** [comment]
Linking to a previous discussion at #4452.

**Comment by DanielRosenwasser** [comment]
@Lenne231 is there a specific scenario you were hit by this?

**Comment by Lenne231** [comment]
Not a specific scenario, i was just playing around with union types.
